platform/security/servlet/security-servlet-session-expiry/src/main/java/org/codice/ddf/security/servlet/expiry/SessionManagementService.java

```diff
@@ -0,0 +1,156 @@
+                            .getTime() - o2.getNotOnOrAfter()
+                            .getTime();
+                    if (l > 0) {
+                        return 1;
+                    } else if (l < 0) {
+                        return -1;
+                    } else {
+                        return 0;
+                    }
+                });
+                timeLeft = values.isEmpty() ?
+                        0 :
+                        values.get(0)
+                                .getNotOnOrAfter()
+                                .getTime();
+            }
+        }
+        return Response.ok(new ByteArrayInputStream(Long.toString(Math.max(
+                timeLeft - clock.millis(), 0))
+                .getBytes(StandardCharsets.UTF_8)))
+                .build();
+    }
+
+    @GET
+    @Path("/renew")
+    public Response getRenewal(@Context HttpServletRequest request) {
+        HttpSession session = request.getSession(false);
+        final Response[] response = new Response[1];
+        if (session != null) {
+            Object securityToken = session.getAttribute(SecurityConstants.SAML_ASSERTION);
+            if (securityToken instanceof SecurityTokenHolder) {
+                SecurityTokenHolder tokenHolder = (SecurityTokenHolder) securityToken;
+                Map<String, SecurityToken> realmTokenMap = tokenHolder.getRealmTokenMap();
+                realmTokenMap.keySet()
+                        .forEach(s -> {
+                            try {
+                                doRenew(s, realmTokenMap.get(s), tokenHolder);
+                            } catch (SecurityServiceException e) {
+                                response[0] = Response.serverError()
+                                        .build();
+                                LOGGER.error("Failed to renew", e);
+                            }
+                        });
+            }
+        }
+        if (response[0] == null) {
+            response[0] = Response.ok()
+                    .build();
+        }
+        return response[0];
+    }
+
+    @GET
+    @Path("/invalidate")
+    public Response getInvalidate(@Context HttpServletRequest request) {
+        StringBuffer requestURL = request.getRequestURL();
+        String requestQueryString = request.getQueryString();
+        return Response.temporaryRedirect(URI.create(requestURL.substring(0,
+                requestURL.indexOf("services"))
+                .concat("logout?noPrompt=true")
+                .concat(requestQueryString != null ? "&" + requestQueryString : "")))
+                .build();
+    }
+
+    private void doRenew(String realm, SecurityToken securityToken, SecurityTokenHolder tokenHolder)
+            throws SecurityServiceException {
+        SAMLAuthenticationToken samlToken =
+                new SAMLAuthenticationToken(securityToken.getPrincipal(), securityToken, realm);
+        Subject subject = securityManager.getSubject(samlToken);
```

Comment:
We probably need a ticket to enhance the code behind this to handle the IdP case correctly. Right now this will work fine if DDF is serving as the IdP. However, if we're using some other IdP, we will most likely not have direct access to their STS and this renewal will fail. If the assertion we're working with is clearly from an IdP, then we would need to use one of the SSO profile bindings (like the SOAP one) that can be done without a user in the loop, to renew the assertion or request a new one.

Reply:
Agreed. Created https://codice.atlassian.net/browse/DDF-3246 for this issue.